### DISCUSSION TITLE
Automatic panel update

### DISCUSSION THREAD

**Discussion by shmuz** [elfmz/far2l#1278] [Q&A]
В настройках панели есть установка, но от неё, похоже, ничего не зависит: панели самостоятельно не обновляются, пока например не нажмёшь `CtrlR`.
```
[ ] Disable automatic update of panels
    if object count exceeds _____
```
Помещаю сторонней программой в папку файл, а его не видно.
Может есть какая-то ещё настройка, чтобы панели обновлялись?

**Comment by unxed** [comment]
У меня видно, но с существенной задержкой иногда. Не рыл условия, от которых это зависит, пока

**Comment by shmuz** [comment]
Ну вот скопировали сторонней программой файл в папку, и ничего не делаем, ждём пока файл появится. Он у вас появляется через какое время, если ничего не делать?

**Reply by shmuz** [reply]
Кажется уловил закономерность: это у меня происходит в папке на смонтированном диске FAT32. А на родной линуксовской файл-системе файл появляется практически сразу.

**Reply by unxed** [reply]
на far32 не тестил. на ext4 появляется, но иногда с задержкой в несколько секунд

**Comment by elfmz** [comment]
Там есть автообновление, работает оно не волшебством, а юзая inotify. Если какаято FS его не поддерживает то работать соответственно не будет. можно попробовать руками поюзать inotify командой https://linux.die.net/man/1/inotifywait чтоб убедится что это не проблема far2l (а вдруг таки проблема, кто знает..)

**Reply by shmuz** [reply]
Проверил: `inotifywait` не детектирует  добавляемый файл на этой файловой системе.
Но возможно есть другие средства, помимо `inotify` ? Фар для Windows вроде этой проблемы не имеет.

**Comment by elfmz** [comment]
фар для виндовс юзает FindFirstChangeNotify&co прямым аналогом чего в линуксе является inotify*, но в линуксе в отличии от винды всегда все чтото гдето немножечко подглючивает. Хотя похоже что последние версии виндов активно внедряют эту передовую функциональность.
В принципе можно добавить периодическую проверку таймстампа и надеятся что оно не будет сильно жрать цпу...

*еще есть fanotify, но во-первых он требует рутовские привилегии, во-вторых сомневаюсь что он будет работать там где inotify не работает, хотя поэкспериментировать конечно можно - тут есть пример https://man7.org/linux/man-pages/man7/fanotify.7.html

**Reply by shmuz** [reply]
> В принципе можно добавить периодическую проверку таймстампа и надеятся что оно не будет сильно жрать цпу...

На FAT32 вроде бы таймстамп папки не меняется при изменениях содержимого.
Я думал сделать в плагине принудительное обновление панелей раз в секунду, но пока ещё не уверен, что это хорошо.

**Reply by elfmz** [reply]
разве? я попробовал делать stat . - у меня меняется если скопировать в текущий дир какой нить файл

**Reply by elfmz** [reply]
принудительное обновление точно не хорошо, если в папке 100500 файлов или папка - подмонированная сетевая фс, то будет тупить

**Reply by shmuz** [reply]
> я попробовал делать stat . - у меня меняется если скопировать в текущий дир какой нить файл

stat я не делал, но far2l не показывает изменения времени модификации даже после принудительного обновления панелей.
(Речь идёт всё о том же подмонтированном FAT32-диске).

**Reply by elfmz** [reply]
А файл добавлялся в директорию или редактировался существующий? изменение существующего таймстамп директории не меняет, меняет только добавление/удаление файла.

**Reply by shmuz** [reply]
Именно добавлялся (копировался из соседней панели прямо в far2l). Время модификации папки назначения не меняется.
Но это и на виндоуз так для фат32.

**Reply by elfmz** [reply]
у меня кстати похоже inotify работает норм для fat32, возможно дело в опциях монтирования. Сделайте `cat /proc/mounts | grep /media/user/boot`  где вместо /media/user/boot подставьте путь где подмонтирована ваша фат32

**Reply by elfmz** [reply]
у меня выглядит так: `/dev/mmcblk0p1 /media/user/boot vfat rw,nosuid,nodev,relatime,uid=1000,gid=1000,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,showexec,utf8,flush,errors=remount-ro 0 0`

**Reply by shmuz** [reply]
Пока что затрудняюсь найти этот путь для подмонтирования. Дело в том, что я работаю через VirtualBox, и при настройке "shared folder" оставил поле "mount point" пустым (тогда он решает сам).

**Reply by elfmz** [reply]
а, так это еще shared folder виртуалки? ну там любые чудеса возможны..

**Reply by shmuz** [reply]
Мало того, этот диск и на винде виртуальный.
Ну фиг с ним пока что, эта проблема у меня сейчас только одна из многих.
